x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AssignmentPlan.java

```diff
@@ -0,0 +1,295 @@
+                .collect(Collectors.toList())) {
+                if (modelAllocations.v2() > 0) {
+                    msg.append(" ");
+                    msg.append(modelAllocations.v1().id());
+                    msg.append(" (mem = ");
+                    msg.append(ByteSizeValue.ofBytes(modelAllocations.v1().memoryBytes()));
+                    msg.append(")");
+                    msg.append(" (allocations = ");
+                    msg.append(modelAllocations.v2());
+                    msg.append("/");
+                    msg.append(modelAllocations.v1().allocations());
+                    msg.append(")");
+                    msg.append(" (threads_per_allocation = ");
+                    msg.append(modelAllocations.v1().threadsPerAllocation());
+                    msg.append(")");
+                }
+            }
+            if (i < nodes.size() - 1) {
+                msg.append('\n');
+            }
+        }
+        return msg.toString();
+    }
+
+    public static Builder builder(Collection<Node> nodes, Collection<Model> models) {
+        return new Builder(nodes, models);
+    }
+
+    static class Builder {
+
+        private final Map<Model, Map<Node, Integer>> assignments;
+        private final Map<Node, Long> remainingNodeMemory;
+        private final Map<Node, Integer> remainingNodeCores;
+        private final Map<Model, Integer> remainingModelAllocations;
+
+        private Builder(Collection<Node> nodes, Collection<Model> models) {
+            if (nodes.stream().collect(Collectors.toSet()).size() != nodes.size()) {
+                throw new IllegalArgumentException("there should be no duplicate nodes");
+            }
+            if (models.stream().collect(Collectors.toSet()).size() != models.size()) {
+                throw new IllegalArgumentException("there should be no duplicate models");
+            }
+
+            assignments = new HashMap<>();
+            remainingNodeMemory = new HashMap<>();
+            remainingNodeCores = new HashMap<>();
+            remainingModelAllocations = new HashMap<>();
+
+            for (Model m : models) {
+                Map<Node, Integer> nodeAssignments = new HashMap<>();
+                for (Node n : nodes) {
+                    nodeAssignments.put(n, 0);
+                    remainingNodeMemory.put(n, n.availableMemoryBytes());
+                    remainingNodeCores.put(n, n.cores());
+                }
+                assignments.put(m, nodeAssignments);
+                remainingModelAllocations.put(m, m.allocations());
+            }
+        }
+
+        int getRemainingCores(Node n) {
+            return remainingNodeCores.get(n);
+        }
+
+        long getRemainingMemory(Node n) {
+            return remainingNodeMemory.get(n);
+        }
+
+        int getRemainingThreads(Model m) {
+            return remainingModelAllocations.get(m) * m.threadsPerAllocation();
+        }
+
+        int getRemainingAllocations(Model m) {
```

Comment:
All these are unboxing and could easily through NPE if something provided a model that was not added to the hashtable (same goes for the node accessors). 

They are package private, which is good. This just all feels very brittle.

Reply:
These are methods on the `Builder` which should have been constructed with the list of nodes/models that are used. Passing a model/node that wasn't part of the collections used for constructing the builder would be a programmatic error.